internal/exec/util/file.go

```diff
@@ -136,7 +136,8 @@ func (u Util) WriteLink(s types.Link) error {
 	}
 
 	if s.Hard {
-		return os.Link(s.Target, path)
+		targetPath := u.JoinPath(s.Target)
```

Comment:
The `Util` used here is initialized in [`createEntries` in the files stage](https://github.com/coreos/ignition/blob/ed13fa92d38e5b0c0f9560e1cdf5f67d0250a88b/internal/exec/stages/files/files.go#L306), and its destination dir is the mount point for this filesystem. This `Util` is then [passed into `create` on all `filesystemEntries`](https://github.com/coreos/ignition/blob/ed13fa92d38e5b0c0f9560e1cdf5f67d0250a88b/internal/exec/stages/files/files.go#L313) (note the `files` variable name is a tad misleading there), `create` then [uses this `Util` to call `WriteLink`](https://github.com/coreos/ignition/blob/ed13fa92d38e5b0c0f9560e1cdf5f67d0250a88b/internal/exec/stages/files/files.go#L207). 

So yes, the `Util` instance that `WriteLink` is called on has its `DestDir` set to the mount point for the filesystem the link is being created on, not the root filesystem mount point. It should not always hardlink to the root partition.